components/brave_wallet/browser/brave_wallet_service.cc

```diff
@@ -5,7 +5,84 @@
 
 #include "brave/components/brave_wallet/browser/brave_wallet_service.h"
 
+#include <utility>
+#include <vector>
+
+#include "base/strings/string_util.h"
+#include "base/values.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/pref_names.h"
 #include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+
+// kBraveWalletUserAssets
+// {
+//    "mainnet": {  // network_id
+//      [
+//        {
+//          "contract_address": "eth",
+//          "name": "Ethereum",
+//          "symbol": "ETH",
+//          "is_erc20": false,
+//          "is_erc721": false,
+//          "decimals": 18,
+//          "visible": true
+//          ...
+//        },
+//        {
+//          "contract_address": "0x0D8775F648430679A709E98d2b0Cb6250d2887EF",
+//          "name": "Basic Attention Token",
+//          "symbol": "BAT",
+//          "is_erc20": true,
+//          "is_erc721": false,
+//          "decimals": 18,
+//          "visible": true
+//          ...
+//        },
+//        {
+//          "contract_address": "0x4729c2017edD1BaDf768595378c668955b537197",
+//          "name": "MOR",
+//          "symbol": "MOR",
+//          "is_erc20": true,
+//          "is_erc721": false,
+//          "decimals": 18,
+//          "visible": true
+//          ...
+//        },
+//        ...
+//      ]
+//    },
+//    "rinkeby": {
+//    ...
+//    },
+//    ...
+//    }
+// }
+//
+//
+namespace {
+
+base::CheckedContiguousIterator<base::Value> FindAsset(
+    base::Value* user_assets_dict,
+    const std::string& contract_address) {
+  DCHECK(user_assets_dict && user_assets_dict->is_list() &&
+         !contract_address.empty());
+
+  auto iter = std::find_if(
+      user_assets_dict->GetList().begin(), user_assets_dict->GetList().end(),
+      [&](const base::Value& value) {
+        if (!value.is_dict()) {
+          return false;
+        }
+        const std::string* address = value.FindStringKey("contract_address");
+        return address &&
+               base::EqualsCaseInsensitiveASCII(*address, contract_address);
```

Comment:
is this intentional in case non-checksum address get added?

Reply:
Was using case-insensitive since user can input address for custom tokens.
Will change the code to validate contract_address using EthAddress and always save checksum address in our pref.

Reply:
Please check https://github.com/brave/brave-core/pull/9969/commits/1300def75f2d3ba4ca31a958e3837b81f69fb0a5.